File: apps/voicememory_mobile/lib/features/curiosity_loop/services/curiosity_adaptive_timing_engine.dart
import '../../../models/journal_entry.dart';

/// Computes when to fire the next curiosity notification from journal rhythm.
class CuriosityAdaptiveTimingEngine {
  const CuriosityAdaptiveTimingEngine();

  static const fallbackDelay = Duration(hours: 24);
  static const minHistoryEntries = 3;
  static const maxSampleSize = 5;

  /// Returns how long to wait after [currentEntryTime] before reminding.
  ///
  /// Uses the average recording hour from the last five entries and targets
  /// the next calendar day at that time. With fewer than three entries,
  /// returns a fixed twenty-four hour delay.
  Duration calculateOptimalDelay({
    required List<JournalEntry> history,
    required DateTime currentEntryTime,
  }) {
    if (history.length < minHistoryEntries) {
      return fallbackDelay;
    }

    final recent = _recentEntries(history, maxSampleSize);
    final averageHour = recent
            .map(_hourOfDay)
            .reduce((sum, hour) => sum + hour) /
        recent.length;

    final currentLocal = currentEntryTime.toLocal();
    final nextDay = DateTime(
      currentLocal.year,
      currentLocal.month,
      currentLocal.day,
    ).add(const Duration(days: 1));

    final targetHour = averageHour.floor().clamp(0, 23);
    final targetMinute =
        ((averageHour - targetHour) * 60).round().clamp(0, 59);

    final target = DateTime(
      nextDay.year,
      nextDay.month,
      nextDay.day,
      targetHour,
      targetMinute,
    );

    final delay = target.difference(currentLocal);
    if (!delay.isNegative) return delay;

    // Safety fallback — next-day target should always be in the future.
    return fallbackDelay;
  }

  static List<JournalEntry> _recentEntries(
    List<JournalEntry> history,
    int limit,
  ) {
    final sorted = [...history]..sort((a, b) => a.createdAt.compareTo(b.createdAt));
    if (sorted.length <= limit) return sorted;
    return sorted.sublist(sorted.length - limit);
  }

  static double _hourOfDay(JournalEntry entry) {
    final local = entry.createdAt.toLocal();
    return local.hour + (local.minute / 60.0);
  }
}

File: apps/api/src/services/loops/curiosity_adaptive_timing_engine.ts
import "server-only";

import type { CuriosityJournalEntryTiming } from "./types";

const FALLBACK_DELAY_MS = 24 * 60 * 60 * 1000;
const MIN_HISTORY_ENTRIES = 3;
const MAX_SAMPLE_SIZE = 5;

/**
 * Computes when to fire the next curiosity notification from journal rhythm.
 */
export class CuriosityAdaptiveTimingEngine {
  calculateOptimalDelayMs(input: {
    history: readonly CuriosityJournalEntryTiming[];
    currentEntryTime: Date;
  }): number {
    if (input.history.length < MIN_HISTORY_ENTRIES) {
      return FALLBACK_DELAY_MS;
    }

    const recent = recentEntries(input.history, MAX_SAMPLE_SIZE);
    const averageHour =
      recent.map(hourOfDay).reduce((sum, hour) => sum + hour, 0) / recent.length;

    const currentLocal = input.currentEntryTime;
    const nextDay = new Date(
      currentLocal.getFullYear(),
      currentLocal.getMonth(),
      currentLocal.getDate() + 1,
    );

    const targetHour = Math.min(23, Math.max(0, Math.floor(averageHour)));
    const targetMinute = Math.min(
      59,
      Math.max(0, Math.round((averageHour - targetHour) * 60)),
    );

    const target = new Date(
      nextDay.getFullYear(),
      nextDay.getMonth(),
      nextDay.getDate(),
      targetHour,
      targetMinute,
    );

    const delay = target.getTime() - currentLocal.getTime();
    return delay > 0 ? delay : FALLBACK_DELAY_MS;
  }
}

function recentEntries(
  history: readonly CuriosityJournalEntryTiming[],
  limit: number,
): CuriosityJournalEntryTiming[] {
  const sorted = [...history].sort(
    (left, right) => Date.parse(left.createdAt) - Date.parse(right.createdAt),
  );
  if (sorted.length <= limit) return sorted;
  return sorted.slice(sorted.length - limit);
}

function hourOfDay(entry: CuriosityJournalEntryTiming): number {
  const local = new Date(entry.createdAt);
  return local.getHours() + local.getMinutes() / 60;
}

export const curiosityAdaptiveTimingEngine = new CuriosityAdaptiveTimingEngine();
